### ISSUE TITLE
optim优化器模型，提高优化效率

### ISSUE THREAD

**Issue by SevenPLF** [PaddlePaddle/Paddle#33950]
在优化器Adam的基础上，做出修改。可以提升优化效率。

参数更新公式 W = W + lr * g。其中lr学习率是人工设定的。在一些新建模型中lr的设定直接影响到了模型训练效率。lr过大会使模型训练结果发散，lr过小又会使模型训练速度过慢。 通过每次梯度更新时，引入loss。由自动化控制理论中的PID调节法，对lr实现动态调整，能使模型训练速度加快。
该方法通过传入的loss，对lr进行评判，从而增加或者减少本次梯度更新的参数。同g共同作用。

![lr-0_001](https://user-images.githubusercontent.com/50135313/124377385-4695b980-dcde-11eb-891a-321114d0a938.png)

**Comment by pangyoki** [comment]
你好，请问是希望paddle提供的Adam API实现这种PID方法吗？能提供相关的论文链接吗？

**Comment by SevenPLF** [comment]
我是自动化专业的，PID调节是自动化领域中对稳态调节的常用的方法，对于不同系统的扰动，通过系统的比例、积分、微分环节的调节，来使得系统快速达到稳态。   我把PID的离散模式应用到了训练的调节器，对学习率使用PID做出评定。在回归方面的预测结果表现的不错。 但是我没有具体的论文。
`       
    pidd = LR * 7.07
    pidi = 0.04 / LR
    opt_PID = torch.optim.PID(net_PID.parameters(), k=LR, pidd=pidd, pidi=pidi, piddFlag=True, pidiFlag=True)



 
    def __init__(self, params, rate_decay=0.01, pidd=0, pidi=0.0, piddFlag=False, pidiFlag=False, k = 0.1, betas=(0.9, 0.999), eps=1e-8,  weight_decay=0, amsgrad=False):



if closure is not None:
            loss = closure.data
            group = self.param_groups[0]
            group['iList'].insert(0, loss)
            #  比例 Kp * L(k)
            self.p = math.tanh(group['k'] * loss)
            #  微分 Kd * [L(k) - L(k - 1)]
            if len(group['iList']) >=2 and group['piddFlag']:
                self.d = math.tanh(group['pidd'] * (group['iList'][0] - group['iList'][1]))
            else:
                self.d = 0
            if abs(abs(self.d) - group['k']) < group['k'] * 0.04:
                group['pidiFlag'] = False
            else:
                group['pidiFlag'] = True
            #  积分 Ki * +Kd
            if group['pidiFlag']:
                if group['pidi'] <=1:
                    group['pidi'] = 1 / pow(group['pidi'],2)
                self.i = math.tanh((1 / group['pidi']) * sum(group['iList']))
            else:
                self.i = 0

            if 'MseLossBackward' in str(closure.grad_fn):
                group['learning_rate'] = abs(self.p + self.i + self.d)
            else:
                self.a = math.tanh(self.p + self.i + self.d)
                group['learning_rate'] = group['k'] * (1 + self.a)

            if len(group['iList']) >=5:
                group['iList'] = group['iList'][0:4]`
这是在Adam step 方法中加了一段代码，计算本次的学习率，在最后更新时候使用此学习率

**Comment by SevenPLF** [comment]
因为是刚开始学习深度学习方面的内容，所以自己手里能验证的模型不多，也没有太多的资源验证。 目前发现在损失函数是MSELoss的回归任务中表现不错。但是限于资源有限，所以想在这里提出自己的想法，希望大家一起来看看。

我做这个是对学习率做出评价。所以不只是adam。对于其他的优化器也应该有提升效率。

**Comment by SevenPLF** [comment]
> 你好，请问是希望paddle提供的Adam API实现这种PID吗？能提供相关的论文链接吗？

麻烦看下我的回复
